internal/zinc-core/src/main/scala/sbt/internal/inc/Locate.scala

```diff
@@ -50,32 +51,38 @@ object Locate {
       className => if (defClass(className)) getF(className).toRight(true) else Left(false)
     }
 
-  def definesClass(entry: File): String => Boolean =
+  def definesClass(entry: File): DefinesClass =
     if (entry.isDirectory)
-      directoryDefinesClass(entry)
+      new DirectoryDefinesClass(entry)
     else if (entry.exists && classpath.ClasspathUtilities.isArchive(entry, contentFallback = true))
-      jarDefinesClass(entry)
+      new JarDefinesClass(entry)
     else
-      const(false)
+      FalseDefinesClass
 
-  def jarDefinesClass(entry: File): String => Boolean =
-    {
-      import collection.JavaConversions._
-      val jar = try { new ZipFile(entry, ZipFile.OPEN_READ) } catch {
-        // ZipException doesn't include the file name :(
-        case e: ZipException => throw new RuntimeException("Error opening zip file: " + entry.getName, e)
-      }
-      val entries = try { jar.entries.map(e => toClassName(e.getName)).toSet } finally { jar.close() }
-      entries.contains _
+  private object FalseDefinesClass extends DefinesClass {
+    override def apply(binaryClassName: String): Boolean = false
+  }
+
+  private class JarDefinesClass(entry: File) extends DefinesClass {
+    import collection.JavaConversions._
+    private val jar = try { new ZipFile(entry, ZipFile.OPEN_READ) } catch {
```

Comment:
AFAICT, this will unnecessarily hold a reference to the jar entry in memory.

Maybe:
```
private val entries = {
  val jar = ???
  try { jar.entries... }
}